File: solutions/go/robot-name/1/robot_name.go
package robotname

import (
	"errors"
	"math/rand"
)

type Robot struct {
	name string
}

const maxPossible = 26 * 26 * 10 * 10 * 10
const letSet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
const numSet = "0123456789"

var namesGenerated = map[string]bool{}

func (r *Robot) Name() (string, error) {
	if len(namesGenerated) == maxPossible {
		return "", errors.New("all named have already been generated")
	}

	if r.name != "" {
		return r.name, nil
	}

	r.name = genName()
	for namesGenerated[r.name] {
		r.name = genName()
	}

	namesGenerated[r.name] = true

	return r.name, nil
}

func (r *Robot) Reset() {
	r.name = ""
}

func genName() string {
	name := make([]byte, 5)
	for idx := range name {
		if idx < 2 {
			name[idx] = letSet[rand.Intn(len(letSet))]
		} else {
			name[idx] = numSet[rand.Intn(len(numSet))]
		}
	}
	return string(name)
}

File: solutions/python/robot-name/1/robot_name.py
from string import ascii_uppercase
from random import shuffle


class Robot:
    names = []

    for c1 in ascii_uppercase:
        for c2 in ascii_uppercase:
            for num in range(1000):
                names.append(f"{c1}{c2}{num}")
    shuffle(names)

    def __init__(self):
        self.name = Robot.names.pop()

    def name(self):
        return self.name

    def reset(self):
        self.name = Robot.names.pop()
